Python_Backend/AnswerGenerator.py

```diff
@@ -28,62 +37,190 @@ def clear_queue(self):
+
+    # Tuning Variables
+    NUM_OF_UP_DOWN_PATTERNS_FOR_YES   = 3
+    NUM_OF_LEFT_RIGHT_PATTERNS_FOR_NO = 3
 
+    FRAME_QUEUE_SIZE = 30
+    STATE_QUEUE_SIZE = max(NUM_OF_LEFT_RIGHT_PATTERNS_FOR_NO, NUM_OF_UP_DOWN_PATTERNS_FOR_YES) * 2
 
+    # Required number of frames in __past_frames queue for state to be added to __past_states queue
+    FRAMES_FOR_STATE_LIST = [25,  # eyes_up
+                             25,  # eyes_down
+                             25,  # eyes_left
+                             25,  # eyes_right
+                             25]  # eyes_center
 
-class IrisAnswerGenerator(AnswerGeneratorInterface):
+    INVALID_BUFFER       = FRAME_QUEUE_SIZE * 2 + 1 # Number of frames queue can be invalid (no state) before an invalid state is added to __past_states
+    MAX_FRAMES_PER_STATE = FRAME_QUEUE_SIZE * 6     # Number of frames allowed in one state, before adding state to __past_states queue again
+
+    INIT_STATE = "INIT"
+
+    __frames_in_invalid_state = 0
+    __frames_in_current_state = 0        
+
+    __past_frames = []
+    __past_states = []
+
+    __state_counters = []  # [eyes_left_counter, eyes_right_counter, eyes_up_counter, eyes_down_counter, eyes_center_counter]
 
-    QUEUE_SIZE = 10
-    __past_states = [QUEUE_SIZE]
 
- 
     def __init__(self):
-        raise NotImplementedError
+        
+       self.clear_queue()
+
 
     def clear_queue(self):
-        """Removes all elements from the queue"""
-        raise NotImplementedError
+        """ Reinitializes queues and clears member variables """
 
+        self.__frames_in_invalid_state = 0
+        self.__frames_in_current_state = 0
 
-    def add_state_to_queue(self):
-        """Removes all elements from the queue"""
-        raise NotImplementedError
+        self.__past_frames.clear()
+        self.__past_states.clear()
+        self.__state_counters.clear()
+
+        for x in range(len(ModelStates.IRIS_STATES)):
+            self.__state_counters.append(0)
+        
+        for x in range(self.FRAME_QUEUE_SIZE):
+            self.__past_frames.append(self.INIT_STATE)
+
+        for x in range(self.STATE_QUEUE_SIZE):
+            self.__past_states.append(self.INIT_STATE)
+
+
+    def add_frame_to_queue(self, frame):
+        """ Adds an element to the queue """
+
+        for face_frame in ModelStates.IRIS_STATES:
+
+            if face_frame == frame:
+                popped_frame = self.__past_frames.pop(0)
+                self.__past_frames.append(frame)
+
+                self.__update_counters(popped_frame, frame)
+                self.__update_states_queue()
+                return
+        
+        raise Exception("IrisAnswerGenerator: Invalid frame cannot be added to queue")
+
+
+    def __update_counters(self, popped_frame, added_frame):
+        """ Increment and decrement counters corresponding to the frame added or removed from __past_frames queue """
+        
+        for x in range(len(ModelStates.IRIS_STATES)):

```

Comment:
Do we need to loop through? Can we just index by the enum? i.e. if popped frame is index 5 instead of looping through we can just:
```
self.__state_counters[popped_frame] -= 1
self.__state_counters[added_frame] += 1
```

Reply:
popped_frame can be INVALID_INIT: 5 so it would break, because state counters only has 5 elements:
__state_counters = []  # [eyes_left_counter, eyes_right_counter, eyes_up_counter, eyes_down_counter, eyes_center_counter]